pkg/securitycontext/admission/admission.go

```diff
@@ -0,0 +1,284 @@
+package admission
+
+import (
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+
+	kadmission "github.com/GoogleCloudPlatform/kubernetes/pkg/admission"
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	scc "github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/serviceaccount"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+
+	allocator "github.com/openshift/origin/pkg/security/controller"
+)
+
+const (
+	validatedSCCAnnotation string = "openshift.io/scc"
+)
+
+func init() {
+	kadmission.RegisterPlugin("SecurityContextConstraint", func(client client.Interface, config io.Reader) (kadmission.Interface, error) {
+		return NewConstraint(client), nil
+	})
+}
+
+type constraint struct {
+	*kadmission.Handler
+	client client.Interface
+	store  cache.Store
+}
+
+var _ kadmission.Interface = &constraint{}
+
+func NewConstraint(kclient client.Interface) kadmission.Interface {
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return kclient.SecurityContextConstraints().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+				return kclient.SecurityContextConstraints().Watch(labels.Everything(), fields.Everything(), resourceVersion)
+			},
+		},
+		&kapi.SecurityContextConstraints{},
+		store,
+		0,
+	)
+	reflector.Run()
+
+	return &constraint{
+		Handler: kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		client:  kclient,
+		store:   store,
+	}
+}
+
+func (c *constraint) Admit(a kadmission.Attributes) error {
+	// 1.  find all SCCs the pod has access to:
+	// 2.  Fully resolve each SCC
+	//     1.  determine uid from the SCC's run as user policy
+	//     2.  set the uid on the SCC
+	// 3.  Match the pod's SC to an SCC; for each SCC:
+	//     1.  generate the SC
+	//     2.  validate against the pod's SC
+
+	if a.GetResource() != string(kapi.ResourcePods) {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(*kapi.Pod)
+	if !ok {
+		return errors.NewBadRequest("a pod was received, but could not convert the request object.")
+	}
+
+	// get the namespace
+	namespace, err := c.client.Namespaces().Get(a.GetNamespace())
```

Comment:
move this check later... the service account and scc fetches can be done with just the name. if you're not working from caches, we want to be as efficient as possible, since this adds several requests to every pod creation